### PR TITLE
memory access sanity checks: abort instead of panic

### PR DESCRIPTION
**Pull request by RalfJung**
Suggested by @Mark-Simulacrum, this should help reduce the performance impact of these checks.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
To be clear, I'm not sure :) I think my measurements were inconclusive. But, in theory, I would *expect* this to be better able to inline etc.

Ideally we'd run perf.r-l.o against this but to do that we'd need to build and publish a debug assertions compiler to compare against...

**Comment by RalfJung** [comment]
> But, in theory, I would expect this to be better able to inline etc.

Is that enough to land this? I don't think I'll have time to do proper benchmarking any time soon... and this code is UB if we don't `abort`, so not having perfect panic-based stacktraces seems acceptable.

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -2057,9 +2057,14 @@ pub unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
         fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
     }
 
-    debug_assert!(is_aligned_and_not_null(src), "attempt to copy from unaligned or null pointer");
-    debug_assert!(is_aligned_and_not_null(dst), "attempt to copy to unaligned or null pointer");
-    debug_assert!(is_nonoverlapping(src, dst, count), "attempt to copy to overlapping memory");
+    if cfg!(debug_assertions)
+        && !(is_aligned_and_not_null(src)
+            && is_aligned_and_not_null(dst)
+            && is_nonoverlapping(src, dst, count))
```
Comment (tmiasko) [inline src/libcore/intrinsics.rs:2063]: What about removing the checks completely when debug assertions are disabled:

```rust
    #[cfg(debug_assertions)]
    if !is_aligned_and_not_null(src) && ...
```

Comment (RalfJung) [reply src/libcore/intrinsics.rs:2063]: We already do that. `if false` is reliably removed.

**Comment by RalfJung** [comment]
@Mark-Simulacrum 
> To be clear, I'm not sure :) I think my measurements were inconclusive. But, in theory, I would expect this to be better able to inline etc.

So how do you suggest we proceed? I don't have time to do proper measurements here. Should I close the PR, or should we land this because it'll surely not degrade debug-std performance?

**Comment by Mark-Simulacrum** [comment]
I can try to find some time to prepare the try commits for perf.rlo to benchmark (by pushing to this PR), hopefully today.

**Comment by Mark-Simulacrum** [comment]
And with reverts of the first two commits..

@bors try @rust-timer queue

I think perf will only pick up the first try build to complete (or maybe neither, unsure) but we can check the other one manually anyway.

**Comment by Mark-Simulacrum** [comment]
@rust-timer build d94127aecf2bae628907d43c376b8fe5070431db

**Comment by Mark-Simulacrum** [comment]
Okay so now both of those are done, the right thing to look at is this comparison: https://perf.rust-lang.org/compare.html?start=ebc6a503e956df6a5e66b34fd34b251b596d3431&end=d94127aecf2bae628907d43c376b8fe5070431db, which compares "debug asserts w/o this PR" and "debug asserts w/ this PR" -- it looks like it's a pretty nice win, 3-10% pretty much across the board. It's also notable that debug assertions are *really* expensive today regardless, 10-30% regressions on many benchmarks.

We probably don't want to apply this pattern exactly across the board -- abort() is after all harder to debug -- but I wonder if we can invent some "low-cost" debug_assert, e.g., one that emulates panic=abort or so (so we still get a backtrace potentially and such, but we still avoid the unwind).

I imagine you probably don't have the bandwidth to explore doing so? Maybe we should re-run benchmarks on a PR that changes `debug_assert!` to literally `abort()` if the condition fails to get a sense of how much performance that can win back.

cc @nnethercote as well

**Comment by RalfJung** [comment]
> I imagine you probably don't have the bandwidth to explore doing so?

I'm afraid I do not. I also have no idea how you did the benchmarking here.^^ But it seems like the PR is a clear win, and only some very low-level methods are changed to not have backtraces any more, so seems like a good sign to me independent of the other changes you suggested.

**Comment by Mark-Simulacrum** [comment]
Okay I'll remove the benchmarking commits on this PR and approve it, I agree it's a sufficient win just by itself.

The way to do benchmarks like this is to get a build from just debug asserts being enabled and another build (critically with same parent on master) with debug asserts plus this PR's changes, then you compare the two try commits directly in perf.rlo.

**Comment by RalfJung** [comment]
Oh and you adjusted CI config to get debug assertions in try builds, I see.

**Comment by Mark-Simulacrum** [comment]
(Should be fine to rollup, as there's no expected impact on non-debug-assert builds).

**Pull request merged**